core/server/common/src/main/java/alluxio/master/StateLockManager.java

```diff
@@ -0,0 +1,188 @@
+import alluxio.conf.PropertyKey;
+import alluxio.conf.ServerConfiguration;
+import alluxio.exception.ExceptionMessage;
+import alluxio.resource.LockResource;
+import alluxio.util.ThreadFactoryUtils;
+
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * Provides graceful and interruptable locking protocol for taking the state lock.
+ *
+ * {@link #lockShared()} will be used by user RPCs and may throw {@link InterruptedException}
+ * based on options passed to {@link #lockExclusive(StateLockOptions)}.
+ *
+ * {@link #lockExclusive(StateLockOptions)} will be used by metadata backups in order to
+ * guarantee paused state during critical tasks.
+ */
+public class StateLockManager {
+  private static final Logger LOG = LoggerFactory.getLogger(StateLockManager.class);
+
+  /** The state-lock */
+  private ReadWriteLock mStateLock = new ReentrantReadWriteLock(true);
+
+  /** The set of threads that are waiting for state-lock in shared mode. */
+  private Set<Thread> mSharedWaiters;
+  /** The set of threads that are holding the state-lock in shared mode. */
+  private Set<Thread> mSharedHolders;
+
+  /** Scheduler that is used for interrupt-cycle. */
+  private ScheduledExecutorService mScheduler;
+  /** The future for the active interrupt cycle */
+  private ScheduledFuture<?> mInterrupterFuture;
+
+  /** Whether state-lock manager is in safe-mode. */
+  private boolean mSafeMode;
+
+  /**
+   * Creates a new state-lock manager.
+   */
+  public StateLockManager() {
+    mSharedWaiters = new ConcurrentHashSet<>();
+    mSharedHolders = new ConcurrentHashSet<>();
+
+    mScheduler = Executors
+        .newSingleThreadScheduledExecutor(ThreadFactoryUtils.build("state-lock-manager-%d", true));
+
+    mSafeMode = ServerConfiguration.getBoolean(PropertyKey.MASTER_BACKUP_STATE_LOCK_SAFE_MODE);
+  }
+
+  /**
+   * Locks the state shared.
+   *
+   * @return the lock resource
+   * @throws InterruptedException
+   */
+  public LockResource lockShared() throws InterruptedException {
+    // Do not allow taking shared lock during safe-mode.
+    if (mSafeMode) {
+      throw new IllegalStateException("Cannot acquire lock shared during safe-mode.");
+    }
+    try {
+      // Register as waiter.
+      mSharedWaiters.add(Thread.currentThread());
+      // Grab the lock interruptibly.
+      mStateLock.readLock().lockInterruptibly();
+    } finally {
+      // Remove itself from waiters.
+      mSharedWaiters.remove(Thread.currentThread());
+    }
+    // Register as holder.
+    mSharedHolders.add(Thread.currentThread());
```

Comment:
hrmmm, this seems tricky and hard to guarantee correct behavior...

Reply:
what is "correct behaviour" ?

Reply:
Any inconsistency could lead to two faulty actions:
1- interrupt a thread when not intended
2- don't interrupt a thread when required

2 is covered by the interrupt-cycle. 
1 can be hardened by proposed locking over sets. Then it can only happen at the end of write-lock and that's acceptable.

Reply:
@gpang I merged waiters and holders sets as we can reliably keep them at sync without heavy locking over the sets. 

In the current state, I don't see any `unacceptable` inconsistency.

- **the interrupt could be missed for a thread:** this will be handled by the next tick.
- **the interrupted thread might already have released the lock:** this will be ignored as a victim of backup. (This can be fixed by keeping track of which threads the manager has interrupted and reset the flags based on that. I don't think that's necessary for the overall scenario though.)

WDYT?